hyperoptimization/workers.py

```diff
@@ -0,0 +1,627 @@
+        :param working_directory: arg used by hpbandster
+        :param args: other args
+        :param kwargs: other keyworded args
+        :return: loss and additional run information
+        """
+
+        # extract S and H if they are hyperparameters
+        self.extract_hypers_from_config(config)
+
+        # define the model
+        model = self.get_external_model(config)
+
+        # setup optimizer
+        if config["optimizer"] == "SGD":
+            optimizer = to.optim.SGD(
+                model.parameters(),
+                lr=config["lr"],
+                momentum=config["sgd_momentum"],
+            )
+        elif config["optimizer"] == "Adam":
+            optimizer = to.optim.Adam(model.parameters(), lr=config["lr"])
+        else:
+            raise NotImplementedError("Currently we support only SGD with momentum and Adam")
+
+        model.to(tvo.get_device())
+        model.device = tvo.get_device()
+
+        # setup TVAE
+        cycliclr_half_step_size = np.ceil(self.N / self.batch_size) * self.epochs_per_half_cycle
+
+        if not self.cyclic_lr:
+            self.min_lr = self.max_lr = config["lr"]
+
+        model = TVAE(
+            external_model=model,
+            shape=None,
+            min_lr=self.min_lr,
+            max_lr=self.max_lr,
+            cycliclr_step_size_up=cycliclr_half_step_size,
+            optimizer=optimizer,
+            precision=to.double,
+        )
+
+        exp_conf = ExpConfig(
+            batch_size=self.batch_size,
+            output=self.output,
+            data_transform=self.data_transform,
+        )
+        estep_conf = self.get_EEM_conf(config)
+
+        # setup training
+        data_fname = self.data_fname
+
+        training = Training(exp_conf, estep_conf, model, data_fname, self.valid_fname)
+        testing = Testing(exp_conf, estep_conf, model, data_fname)
+        print("\nlearning...")
+        training_results = []
+        for train_log in training.run(int(budget)):
+            train_log.print()
+            training_results.append(train_log._results)
+
+        testing_results = []
+        for test_log in testing.run(1):
+            # test_log.print()
+            testing_results.append(test_log._results)
+
+        train_F, subs = get(training_results[-1], "train_F", "train_subs")
+        valid_F, subs = get(training_results[-1], "test_F", "test_subs")
+
+        test_F, subs = get(testing_results[-1], "test_F", "test_subs")
+
+        # optimizable = -train_F  # HpBandSter always minimizes
+        #
+        # if self.valid_fname:
+        #     optimizable = -valid_F  # HpBandSter always minimizes
+
+        return {
+            "loss": -train_F if -train_F else np.nan,
+            "info": {
+                "test accuracy": test_F,
+                "train accuracy": train_F,
+                "validation accuracy": valid_F,
+                "number of parameters": model._external_model.number_of_parameters(),
+            },
+        }
```

Comment:
Do I understand correctly that this implies dependency on a particular model and optimization metric? (similar comment above)

Reply:
This is for a particular worker. Individual models require individual workers.